hazelcast/src/main/java/com/hazelcast/quorum/QuorumServiceImpl.java

```diff
@@ -0,0 +1,290 @@
+        }
+    };
+
+
+    public QuorumServiceImpl(NodeEngineImpl nodeEngine) {
+        this.nodeEngine = nodeEngine;
+        this.eventService = nodeEngine.getEventService();
+        Node node = nodeEngine.getNode();
+        this.quorumConfigs = node.getConfig().getQuorumConfigs();
+        initializeQuorums();
+        initializeListeners();
+    }
+
+    private void initializeQuorums() {
+        for (QuorumConfig quorumConfig : quorumConfigs.values()) {
+            initializeDefaultQuorumResolver(quorumConfig);
+        }
+    }
+
+
+    private void initializeDefaultQuorumResolver(QuorumConfig quorumConfig) {
+        String quorumName = quorumConfig.getName();
+        getOrPutSynchronized(quorumCallbacks, quorumName, quorumCallbacks, cf);
+        int size = quorumConfig.getSize();
+        if (size > 0) {
+            MemberCountQuorumResolver resolver = new MemberCountQuorumResolver(size);
+            quorumResolvers.put(quorumName, resolver);
+        }
+    }
+
+    private void initializeListeners() {
+        for (Map.Entry<String, QuorumConfig> configEntry : quorumConfigs.entrySet()) {
+            QuorumConfig config = configEntry.getValue();
+            for (QuorumListenerConfig listenerConfig : config.getListenerConfigs()) {
+                if (listenerConfig != null) {
+                    QuorumListener listener = null;
+                    if (listenerConfig.getImplementation() != null) {
+                        listener = listenerConfig.getImplementation();
+                    } else if (listenerConfig.getClassName() != null) {
+                        try {
+                            listener = ClassLoaderUtil
+                                    .newInstance(nodeEngine.getConfigClassLoader(), listenerConfig.getClassName());
+                        } catch (Exception e) {
+                            throw ExceptionUtil.rethrow(e);
+                        }
+                    }
+                    if (listener != null) {
+                        String instanceName = configEntry.getKey();
+                        addQuorumListener(instanceName, listener);
+                    }
+                }
+            }
+        }
+    }
+
+    public void addQuorumListener(String name, QuorumListener listener) {
+        eventService.registerLocalListener(SERVICE_NAME, name, listener);
+    }
+
+
+    public void resolveQuorum(Operation op) {
+        if (isNamedOperation(op) && hasQuorumAwareService(op)) {
+            String quorumName = getQuorumName(op);
+            if (quorumName != null && isQuorumNeeded(quorumName, op)) {
+                doQuorum(quorumName);
+            }
+        }
+    }
+
+    private String getQuorumName(Operation op) {
+        QuorumAwareService service = op.getService();
+        String name = ((NamedOperation) op).getName();
+        return service.getQuorumName(name);
+    }
+
+    private boolean hasQuorumAwareService(Operation op) {
```

Comment:
I would rename to 'isQuorumAware'. The fact that it is implemented using the the service is not relevant for the intent of this method.